### PR TITLE
Add the more fusion (conv3d and batchnorm)support in pytorch quantization flow

### PR DESCRIPTION
**Pull request by lly-zero-one**

### CODE REVIEW

File: torch/quantization/fuse_modules.py [hunk 1]
```diff
@@ -45,12 +48,19 @@ def fuse_conv_bn_relu(conv, bn, relu):
...
     assert(conv.training == bn.training == relu.training),\
         "Conv and BN both must be in the same mode (train or eval)."
-
     if conv.training:
-        return torch_fused.ConvBnReLU2d(conv, bn, relu)
+        assert not relu.inplace, 'We only support fusion of non-inplace ReLU.'
+        if isinstance(conv, torch.nn.Conv3d):
+            return torch_fused.ConvBnReLU3d(conv, bn, relu)
+        else:
+            return torch_fused.ConvBnReLU2d(conv, bn, relu)
```
Comment (jerryzh168) [inline torch/quantization/fuse_modules.py]: can you do the `if isinstance(conv, torch.nn.Conv3d):` separately before `if conv.training`?

Also please remove `assert not relu.inplace, 'We only support fusion of non-inplace ReLU.'`

Comment (raghuramank100) [reply torch/quantization/fuse_modules.py]: Yes, inplace ReLU fusion is now supported.

File: torch/quantization/fuse_modules.py [hunk 2]
```diff
@@ -45,12 +48,19 @@ def fuse_conv_bn_relu(conv, bn, relu):
     """
     assert(conv.training == bn.training == relu.training),\
         "Conv and BN both must be in the same mode (train or eval)."
-
     if conv.training:
-        return torch_fused.ConvBnReLU2d(conv, bn, relu)
+        assert not relu.inplace, 'We only support fusion of non-inplace ReLU.'
+        if isinstance(conv, torch.nn.Conv3d):
```
Comment (raghuramank100) [inline torch/quantization/fuse_modules.py]: You also need to add a qat module here for training support. https://github.com/pytorch/pytorch/blob/master/torch/nn/intrinsic/qat/modules/conv_fused.py#L156

Please make sure that conv3d+ bn fusion is supported for both training and eval.

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see comments

File: torch/quantization/fuse_modules.py (hunk 2 above)
Comment (lly-zero-one) [reply torch/quantization/fuse_modules.py]: Let's add it in another PR. This PR is mainly for post-training quantization flow.

File: torch/quantization/fuse_modules.py [hunk 3]
```diff
@@ -120,15 +131,10 @@ def fuse_modules(model, modules_to_fuse, inplace=False, fuser_func=fuse_known_mo
...
-    * conv, bn, relu
-
-    * conv, relu
-
-    * linear, relu
-
+    conv, bn
+    conv, bn, relu
+    conv, relu
+    linear, relu
```
Comment (jerryzh168) [inline torch/quantization/fuse_modules.py]: add `bn, relu` here?

File: torch/nn/intrinsic/modules/fused.py [hunk 4]
```diff
@@ -46,3 +46,40 @@ def __init__(self, conv, bn, relu):
...
+    During quantization this will be replaced with the corresponding fused module."""
+    def __init__(self, conv, bn, relu):
+        assert type(conv) == Conv3d and type(bn) == BatchNorm3d and \
+            type(relu) == ReLU, 'Incorrect types for input modules{}{}{}' \
+            .format(type(conv), type(bn), type(relu))
+        super(ConvBnReLU3d, self).__init__(conv, bn, relu)
+
+
+class BNReLU2d(torch.nn.Sequential):
+    r"""This is a sequential container which calls the Conv 2d and ReLU modules.
```
Comment (jerryzh168) [inline torch/nn/intrinsic/modules/fused.py]: comments need to be updated

Comment (jerryzh168) [reply torch/nn/intrinsic/modules/fused.py]: also `Conv2d` looks better than `Conv 2d` I think

**Review by jerryzh168 — APPROVED** [review]
Looks good, please see some comments inline

File: torch/quantization/fuse_modules.py [hunk 5]
```diff
@@ -45,12 +48,15 @@ def fuse_conv_bn_relu(conv, bn, relu):
     """
     assert(conv.training == bn.training == relu.training),\
         "Conv and BN both must be in the same mode (train or eval)."
-
+    is_3d = isinstance(conv, torch.nn.Conv3d)
```
Comment (jerryzh168) [inline torch/quantization/fuse_modules.py]: actually, do you need to verify Bn and Conv have matching dimensions here?

File: torch/quantization/fuse_modules.py [hunk 6]
```diff
@@ -26,15 +26,18 @@ def fuse_conv_bn(conv, bn):
         assert bn.num_features == conv.out_channels, 'Output channel of Conv2d must match num_features of BatchNorm2d'
         assert bn.affine, 'Only support fusing BatchNorm2d with affine set to True'
         assert bn.track_running_stats, 'Only support fusing BatchNorm2d with tracking_running_stats set to True'
-        return torch.nn.intrinsic.ConvBn2d(conv, bn)
+        if isinstance(conv, torch.nn.Conv3d):
```
Comment (jerryzh168) [inline torch/quantization/fuse_modules.py]: same here

File: torch/quantization/fuse_modules.py (hunk 5 above)
Comment (lly-zero-one) [reply torch/quantization/fuse_modules.py]: I think we need to apply the same constraint as ConvBN.

Comment (jerryzh168) [reply torch/quantization/fuse_modules.py]: you mean ConvBn2d? yeah please add that as well
